# Problem set 3: Text analysis of DOJ press releases

**Total points (without extra credit)**: 52 

- For background:

    - DOJ is the federal law enforcement agency responsible for federal prosecutions; this contrasts with the local prosecutions in the Cook County dataset we analyzed earlier. Here's a short explainer on which crimes get prosecuted federally versus locally: https://www.criminaldefenselawyer.com/resources/criminal-defense/federal-crime/state-vs-federal-crimes.htm#:~:text=Federal%20criminal%20prosecutions%20are%20handled,of%20state%20and%20local%20law. 
    - Here's the Kaggle that contains the data: https://www.kaggle.com/jbencina/department-of-justice-20092018-press-releases 
    - Here's the code the dataset creator used to scrape those press releases here if you're interested: https://github.com/jbencina/dojreleases

## 0.0 Import packages

In [3]:
## helpful packages
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import random
import re
import string

## nltk imports
import nltk
### uncomment and run these lines if you haven't downloaded relevant nltk add-ons yet
### nltk.download('averaged_perceptron_tagger')
### nltk.download('stopwords')
from nltk import pos_tag
from nltk.tokenize import word_tokenize, wordpunct_tokenize
from nltk.stem.snowball import SnowballStemmer
from nltk.corpus import stopwords

## spacy imports
import spacy
### uncomment and run the below line if you haven't loaded the en_core_web_sm library yet
### ! python -m spacy download en_core_web_sm
import en_core_web_sm
nlp = en_core_web_sm.load()

## vectorizer
from sklearn.feature_extraction.text import CountVectorizer

## sentiment
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

## lda
from gensim import corpora
import gensim

## repeated printouts and wide-format text
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
pd.set_option('display.max_colwidth', None)

## 0.1 Load and clean text data

In [5]:
## first, unzip the file pset3_inputdata.zip 
## then, run this code to load the unzipped json file and convert to a dataframe
## (may need to change the pathname depending on where you store stuff)
## and convert some of the attributes from lists to values
doj = pd.read_json("combined.json", lines = True)

## due to json, topics are in a list so remove them and concatenate with ;
doj['topics_clean'] = ["; ".join(topic) 
                      if len(topic) > 0 else "No topic" 
                      for topic in doj.topics]

## similarly with components
doj['components_clean'] = ["; ".join(comp) 
                           if len(comp) > 0 else "No component" 
                           for comp in doj.components]

## drop older columns from data
doj = doj[['id', 'title', 'contents', 'date', 'topics_clean', 
           'components_clean']].copy()

doj.head()

id  \
0     None   
1  12-919    
2  11-1002   
3   10-015   
4   18-898   

                                                                                                          title  \
0                                                                  Convicted Bomb Plotter Sentenced to 30 Years   
1                              $1 Million in Restitution Payments Announced to Preserve North Carolina Wetlands   
2                 $1 Million Settlement Reached for Natural Resource Damages at Superfund Site in Massachusetts   
3                                          10 Las Vegas Men Indicted \r\nfor Falsifying Vehicle Emissions Tests   
4  $100 Million Settlement Will Speed Cleanup Work at Centredale Manor Superfund Site in North Providence, R.I.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                

## 1. Tagging and sentiment scoring (17 points)

Focus on the following press release: `id` == "17-1204" about this pharmaceutical kickback prosecution: https://www.forbes.com/sites/michelatindera/2017/11/16/fentanyl-billionaire-john-kapoor-to-plead-not-guilty-in-opioid-kickback-case/?sh=21b8574d6c6c 

The `contents` column is the one we're treating as a document. You may need to to convert it from a pandas series to a single string.

We'll call the raw string of this press release `pharma`

In [7]:
## your code to subset to one press release and take the string
pharma_focus = doj[doj["id"] == "17-1204"]

pharma = pharma_focus["contents"].values[0]

### 1.1 part of speech tagging (3 points)

A. Preprocess the `pharma` press release to remove all punctuation / digits (you can use `.isalpha()` to subset)

B. With the preprocessed press release from part A, use the part of speech tagger within nltk to tag all the words in that one press release with their part of speech. 

C. Using the output from B, extract the adjectives and sort those adjectives from most occurrences to fewest occurrences. Print a dataframe with the 5 most frequent adjectives and their counts in the `pharma` release. See here for a list of the names of adjectives within nltk: https://pythonprogramming.net/natural-language-toolkit-nltk-part-speech-tagging/

**Resources**:

- Documentation for `.isalpha()`: https://www.w3schools.com/python/ref_string_isalpha.asp

In [9]:
import nltk
nltk.download('punkt_tab')  # downloads to default path

# Show all paths nltk is using
print(nltk.data.path)

[nltk_data] Downloading package punkt_tab to /Users/trak/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

['/Users/trak/nltk_data', '/opt/anaconda3/envs/textenv/nltk_data', '/opt/anaconda3/envs/textenv/share/nltk_data', '/opt/anaconda3/envs/textenv/lib/nltk_data', '/usr/share/nltk_data', '/usr/local/share/nltk_data', '/usr/lib/nltk_data', '/usr/local/lib/nltk_data']


In [10]:
from nltk.tokenize import sent_tokenize, word_tokenize, PunktSentenceTokenizer
from nltk.data import load

# PART A: remove punctuation
tokenizer = load('tokenizers/punkt/english.pickle')
sentences = tokenizer.tokenize(pharma)
pharma_tokens = [word for sent in sentences for word in word_tokenize(sent)]
pharma_alpha = [word.lower() for word in pharma_tokens if word.isalpha()]

# PART B: tag all words
pharma_pos = pos_tag(pharma_alpha)

In [11]:
## your code here for part of speech tagging

# PART C: extract adj and occurences
from collections import Counter
import pandas as pd

adjectives = [word for word, tag in pharma_pos if tag in ('JJ', 'JJR', 'JJS')]
adj_counts = Counter(adjectives)
top_adjs_doj = pd.DataFrame(adj_counts.most_common(5), columns=["Adjective", "Count"])
top_adjs_doj

Adjective  Count
0      former      8
1     special      7
2      opioid      5
3    criminal      5
4  nationwide      4

## 1.2 named entity recognition (4 points)

A. Using the original `pharma` press release (so the one before stripping punctuation/digits), use spaCy to extract all named entities from the press release.

B. Print the unique named entities with the tag: `LAW`

In [14]:
## PART A
import spacy

doc = nlp(pharma)

for ent in doc.ents:
    print(f"{ent.text} ({ent.label_})")

Insys Therapeutics Inc. (ORG)
today (DATE)
Fentanyl (PERSON)
More than 20,000 (CARDINAL)
Americans (NORP)
last year (DATE)
millions (CARDINAL)
Jeff Sessions (PERSON)
This Justice Department (ORG)
Trump (PERSON)
American (NORP)
”John N. Kapoor (PERSON)
74 (DATE)
Phoenix (GPE)
Ariz. (GPE)
the Board of Directors (ORG)
Insys (ORG)
this morning (TIME)
Arizona (GPE)
RICO (LAW)
Kapoor (PERSON)
Executive (ORG)
Board (ORG)
Insys (ORG)
Phoenix (GPE)
today (DATE)
U.S. (GPE)
District Court (ORG)
Boston (GPE)
a later date (DATE)
today (DATE)
Boston (GPE)
Insys (ORG)
December 2016.The (DATE)
Kapoor (GPE)
Michael L. Babich (PERSON)
40 (DATE)
Scottsdale (GPE)
Ariz. (GPE)
Alec Burlakoff (PERSON)
42 (DATE)
Charlotte (GPE)
N.C. (GPE)
Richard M. Simon (PERSON)
46 (DATE)
Seal Beach (GPE)
Calif. (GPE)
Sunrise Lee (PERSON)
36 (DATE)
Bryant City (GPE)
Mich. (GPE)
Joseph A. Rowan (PERSON)
43 (DATE)
Panama City (GPE)
Fla. (GPE)
Managed Markets (ORG)
Michael J. Gurry (PERSON)
53 (DATE)
Scottsdale (GPE)
Ariz. (GP

In [15]:
## your code here for part B

spacy_pharma = nlp(pharma)
law = set([ent.text for ent in spacy_pharma.ents if ent.label_ == "LAW"])
law

{'RICO', 'the Controlled Substances Act'}

C. Use Google to summarize in one sentence what the `RICO` named entity means and why this might apply to a pharmaceutical kickbacks case (and not just a mafia case...) 

## PART C

### RICO stands for the Racketeer Influenced and Corrupt Organizations Act which is a federal law that allows prosecution of individuals involved in organized and long-term criminal activities including executive corporate executives, making it applicable to pharmaceutical kickbacks (bribing companies to prescribe or promote products)

D. You want to extract the possible sentence lengths the CEO is facing; pull out the named entities with (1) the label `DATE` and (2) that contain the word year or years (hint: you may want to use the `re` module for that second part). Print these named entities.

In [19]:
## your code here
pattern = r"(year|years)"
date = [ent.text for ent in spacy_pharma.ents if ent.label_ == "DATE" and re.search(pattern, ent.text)]
date

['last year', '20 years', 'three years', 'five years', 'three years']

E. Pull and print the original parts of the press releases where those year lengths are mentioned (e.g., the sentences or rough region of the press release). Describe in your own words (1 sentence) what length of sentence (prison) and probation (supervised release) the CEO may be facing if convicted after this indictment (if there are multiple lengths mentioned describe the maximum). 

**Hint**: you may want to use re.search or re.findall 

- For part E, you can use `re.search` and `re.findall`, or anything that works 😳.

In [21]:
## your code here
import re
import nltk
from nltk.tokenize import sent_tokenize

sentences_with_years = [sent for sent in spacy_pharma.sents if any("year" in sent.text for year in date)]
sentences_with_years

["More than 20,000 Americans died of synthetic opioid overdoses last year, and millions are addicted to opioids.,
 The charges of conspiracy to commit RICO and conspiracy to commit mail and wire fraud each provide for a sentence of no greater than 20 years in prison, three years of supervised release and a fine of $250,000, or twice the amount of pecuniary gain or loss.  ,
 The charges of conspiracy to violate the Anti-Kickback Law provide for a sentence of no greater than five years in prison, three years of supervised release and a $25,000 fine.]

## 1.3 sentiment analysis  (10 points)

A. Subset the press releases to those labeled with one of three topics via `topics_clean`: Civil Rights, Hate Crimes, and Project Safe Childhood. We'll call this `doj_subset` going forward and it should have 717 rows.



In [23]:
## your code here for subsetting
target_topics = ["Civil Rights", "Hate Crimes", "Project Safe Childhood"]

# Subset the data
doj_subset = doj[doj["topics_clean"].isin(target_topics)]

# Check number of rows
print(f"Subset contains {len(doj_subset)} rows.")
doj_subset.head()

Subset contains 717 rows.


id  \
77   17-1235   
155  15-1522   
157   16-213   
162   16-381   
168   14-464   

                                                                                                                        title  \
77   Additional Former Correctional Officer Pleads Guilty to Beating of Handcuffed and Shackled Inmate at Angola State Prison   
155                                                            Alabama Man Found Guilty of Aggravated Sexual Abuse of a Child   
157                                                        Alabama Man Indicted on Child Pornography and Sex Tourism Charges    
162                                           Alabama Man Indicted for Producing Child Pornography Involving Multiple Victims   
168                                Alabama Man Indicted for Threatening African-American Man and Another Person at Restaurant   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      contents  \
77                                                                                                                                                                                                                                                                                                                                                                                                                                A former supervisory correctional officer at Louisiana State Penitentiary in Angola, Louisiana, pleaded guilty yesterday in connection with the beating of a handcuffed and shackled inmate, in addition to conspiring to cover up their misconduct by falsifying official rec

B. Write a function that takes one press release string as an input and:

- Removes named entities from each press release string (**Hint**: you may want to use `re.sub` with an or condition)
- Scores the sentiment of the entire press release using the `SentimentIntensityAnalyzer` and `polarity_scores`
- Returns the length-four (negative, positive, neutral, compound) sentiment dictionary (any order is fine)

Apply that function to each of the press releases in `doj_subset`. 

**Hints**: 

- A function + list comprehension to execute will takes about 30 seconds on a respectable local machine and about 2 mins on jhub; if it's taking a very long time, you may want to check your code for inefficiencies. If you can't fix those, for partial credit on this part/full credit on remainder, you can take a small random sample of the 717


In [25]:
def func(text):
    sent_obj = SentimentIntensityAnalyzer()
    sub = nlp(text)
    entities = set(ent.text for ent in sub.ents)
    
    cleaned_text = text
    
    for ent in entities:
        pattern = re.escape(ent)  # make it safe
        cleaned_text = re.sub(pattern, '', cleaned_text)

    sentiment = sent_obj.polarity_scores(cleaned_text)
    return sentiment

In [26]:
doj_func = doj_subset.contents.apply(func)

In [162]:
doj_func

doj_subset_sent = pd.DataFrame(doj_func.tolist())
doj_subset_wscore = pd.concat([doj_subset.reset_index(drop=True),doj_subset_sent],axis=1)

doj_subset_wscore.head()


doj_filt = doj_subset_wscore.sort_values("neg",ascending=False)
doj_filt.head(2)
doj_filt[['id','contents','neg']]

77         {'neg': 0.2, 'neu': 0.751, 'pos': 0.049, 'compound': -0.9931}
155      {'neg': 0.134, 'neu': 0.798, 'pos': 0.069, 'compound': -0.9325}
157      {'neg': 0.092, 'neu': 0.832, 'pos': 0.076, 'compound': -0.7579}
162      {'neg': 0.123, 'neu': 0.794, 'pos': 0.083, 'compound': -0.9037}
168      {'neg': 0.178, 'neu': 0.778, 'pos': 0.044, 'compound': -0.9864}
                                      ...                               
13002    {'neg': 0.159, 'neu': 0.778, 'pos': 0.062, 'compound': -0.9737}
13032     {'neg': 0.081, 'neu': 0.823, 'pos': 0.096, 'compound': 0.5267}
13034    {'neg': 0.151, 'neu': 0.757, 'pos': 0.092, 'compound': -0.9571}
13068     {'neg': 0.139, 'neu': 0.735, 'pos': 0.126, 'compound': -0.912}
13081     {'neg': 0.149, 'neu': 0.82, 'pos': 0.031, 'compound': -0.9921}
Name: contents, Length: 717, dtype: object

id  \
0  17-1235   
1  15-1522   
2   16-213   
3   16-381   
4   14-464   

                                                                                                                      title  \
0  Additional Former Correctional Officer Pleads Guilty to Beating of Handcuffed and Shackled Inmate at Angola State Prison   
1                                                            Alabama Man Found Guilty of Aggravated Sexual Abuse of a Child   
2                                                        Alabama Man Indicted on Child Pornography and Sex Tourism Charges    
3                                           Alabama Man Indicted for Producing Child Pornography Involving Multiple Victims   
4                                Alabama Man Indicted for Threatening African-American Man and Another Person at Restaurant   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    contents  \
0                                                                                                                                                                                                                                                                                                                                                                                                                               A former supervisory correctional officer at Louisiana State Penitentiary in Angola, Louisiana, pleaded guilty yesterday in connection with the beating of a handcuffed and shackled inmate, in addition to conspiring to cover up their misconduct by falsifying official records and lying to internal

id  \
13  14-248   
34  13-312   

                                                                                                    title  \
13  Albuquerque Man Charged with Federal Hate Crime Related to Anti-Semitic Threats Against Businesswoman   
34                     Aryan Brother Inmate Sentenced for Federal Hate Crime for Assaulting Fellow Inmate   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 contents  \
13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The Department of Justice announced that this morning John W. Ng, 58, of Albuquerque, N.M., made his initial appearance in federal court on a criminal complaint charging him with a hate crime offense.  This charge is related to anti-Semitic threats Ng made against a Jewish woman who owns and operates the Nosh Jewish Delicatessen and Bakery in Albuquerque. Ng was arrested by the FBI on March 7, 2014, based o

id  \
13    14-248   
34    13-312   
632   16-718   
22    11-626   
594  10-1194   
..       ...   
324   17-132   
392   16-539   
346   17-003   
72    17-271   
551  15-1559   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 contents  \
13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

C. Add the four sentiment scores to the `doj_subset` dataframe to create a dataframe: `doj_subset_wscore`. Sort from highest neg to lowest neg score and print the top `id`, `contents`, and `neg` columns of the two most neg press releases. 

Notes:

- Don't worry if your sentiment score differs slightly from our output on GitHub; differences in preprocessing can lead to diff scores

In [49]:
doj_func

doj_subset_sent = pd.DataFrame(doj_func.tolist())
doj_subset_wscore = pd.concat([doj_subset.reset_index(drop=True),doj_subset_sent],axis=1)
doj_filt = doj_subset_wscore.sort_values("neg",ascending=False)
doj_filt.head(2)
doj_filt[['id','contents','neg']]

77         {'neg': 0.2, 'neu': 0.751, 'pos': 0.049, 'compound': -0.9931}
155      {'neg': 0.134, 'neu': 0.798, 'pos': 0.069, 'compound': -0.9325}
157      {'neg': 0.092, 'neu': 0.832, 'pos': 0.076, 'compound': -0.7579}
162      {'neg': 0.123, 'neu': 0.794, 'pos': 0.083, 'compound': -0.9037}
168      {'neg': 0.178, 'neu': 0.778, 'pos': 0.044, 'compound': -0.9864}
                                      ...                               
13002    {'neg': 0.159, 'neu': 0.778, 'pos': 0.062, 'compound': -0.9737}
13032     {'neg': 0.081, 'neu': 0.823, 'pos': 0.096, 'compound': 0.5267}
13034    {'neg': 0.151, 'neu': 0.757, 'pos': 0.092, 'compound': -0.9571}
13068     {'neg': 0.139, 'neu': 0.735, 'pos': 0.126, 'compound': -0.912}
13081     {'neg': 0.149, 'neu': 0.82, 'pos': 0.031, 'compound': -0.9921}
Name: contents, Length: 717, dtype: object

id  \
13  14-248   
34  13-312   

                                                                                                    title  \
13  Albuquerque Man Charged with Federal Hate Crime Related to Anti-Semitic Threats Against Businesswoman   
34                     Aryan Brother Inmate Sentenced for Federal Hate Crime for Assaulting Fellow Inmate   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 contents  \
13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The Department of Justice announced that this morning John W. Ng, 58, of Albuquerque, N.M., made his initial appearance in federal court on a criminal complaint charging him with a hate crime offense.  This charge is related to anti-Semitic threats Ng made against a Jewish woman who owns and operates the Nosh Jewish Delicatessen and Bakery in Albuquerque. Ng was arrested by the FBI on March 7, 2014, based o

id  \
13    14-248   
34    13-312   
632   16-718   
22    11-626   
594  10-1194   
..       ...   
324   17-132   
392   16-539   
346   17-003   
72    17-271   
551  15-1559   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 contents  \
13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

D. With the dataframe from part C, find the mean compound sentiment score for each of the three topics in `topics_clean` using group_by and agg.

E. Add a 1 sentence interpretation of why we might see the variation in scores (remember that compound is a standardized summary where -1 is most negative; +1 is most positive)


In [53]:
doj_filt.groupby("topics_clean")["compound"].agg("mean")

print("""
Hate Crimes, scoring very negative, are generally described using words with violent undertones while 
Civil Rights, scoring relatively neutral, tend to be described with political words that try to 
express as little bias as possible. Project Safe Childhood could be in between the two due to the writers
describing consequences of an unsafe childhood in combination with the positives of a safe childhood.
""")

topics_clean
Civil Rights             -0.081104
Hate Crimes              -0.933764
Project Safe Childhood   -0.658967
Name: compound, dtype: float64


Hate Crimes, scoring very negative, are generally described using words with violent undertones while 
Civil Rights, scoring relatively neutral, tend to be described with political words that try to 
express as little bias as possible. Project Safe Childhood could be in between the two due to the writers
describing consequences of an unsafe childhood in combination with the positives of a safe childhood.



# 2. Topic modeling (25 points)

For this question, use the `doj_subset_wscores` data that is restricted to civil rights, hate crimes, and project safe childhood and with the sentiment scores added


In [57]:
doj_subset_wscore.head()

id  \
0  17-1235   
1  15-1522   
2   16-213   
3   16-381   
4   14-464   

                                                                                                                      title  \
0  Additional Former Correctional Officer Pleads Guilty to Beating of Handcuffed and Shackled Inmate at Angola State Prison   
1                                                            Alabama Man Found Guilty of Aggravated Sexual Abuse of a Child   
2                                                        Alabama Man Indicted on Child Pornography and Sex Tourism Charges    
3                                           Alabama Man Indicted for Producing Child Pornography Involving Multiple Victims   
4                                Alabama Man Indicted for Threatening African-American Man and Another Person at Restaurant   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    contents  \
0                                                                                                                                                                                                                                                                                                                                                                                                                               A former supervisory correctional officer at Louisiana State Penitentiary in Angola, Louisiana, pleaded guilty yesterday in connection with the beating of a handcuffed and shackled inmate, in addition to conspiring to cover up their misconduct by falsifying official records and lying to internal

## 2.1 Preprocess the data by removing stopwords, punctuation, and non-alpha words (5 points)

A. Write a function that:

- Takes in a single raw string in the `contents` column from that dataframe
- Does the following preprocessing steps:

    - Converts the words to lowercase
    - Removes stopwords, adding the custom stopwords in your code cell below to the default stopwords list
    - Only retains alpha words (so removes digits and punctuation)
    - Only retains words 4 characters or longer
    - Uses the snowball stemmer from nltk to stem

- Returns a joined preprocessed string
    
B. Use `apply` or list comprehension to execute that function and create a new column in the data called `processed_text`
    
C. Print the `id`, `contents`, and `processed_text` columns for the following press releases:

id = 16-718 (this case: https://www.seattletimes.com/nation-world/doj-miami-police-reach-settlement-in-civil-rights-case/)

id = 16-217 (this case: https://www.wlbt.com/story/32275512/three-mississippi-correctional-officers-indicted-for-inmate-assault-and-cover-up/)
    
**Resources**:

- Here's code examples for the snowball stemmer: https://www.geeksforgeeks.org/snowball-stemmer-nlp/

In [60]:
custom_doj_stopwords = ["civil", "rights", "division", "department", "justice",
                        "office", "attorney", "district", "case", "investigation", "assistant",
                       "trial", "assistance", "assist"]

In [62]:
## your code defining a text processing function
from nltk.corpus import stopwords
from nltk.stem.snowball import SnowballStemmer
import re

default_stopwords = set(stopwords.words('english'))
all_stopwords = default_stopwords.union(custom_doj_stopwords)

stemmer = SnowballStemmer("english")

def preprocess_text(text):
    text = text.lower()

    tokens = re.findall(r'\b\w+\b', text)

    processed = [
        stemmer.stem(word) for word in tokens
        if word.isalpha() and word not in all_stopwords and len(word) >= 4
    ]

    return " ".join(processed)

In [64]:
## your code executing the function
doj_subset_wscore["processed_text"] = doj_subset_wscore["contents"].apply(preprocess_text)

In [65]:
## your code showing the examples
example_ids = ["16-718", "16-217"]
doj_subset_wscore.loc[doj_subset_wscore["id"].isin(example_ids), ["id", "contents", "processed_text"]]

id  \
313  16-217   
632  16-718   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       contents  \
313  The Justice Department has reached a comprehensive settlement agreement with the city of Miami and the Miami Police Department (MPD) resolving the Justice Department’s investigation of officer-involved shootings by MPD officers, announced Principal Deputy Assistant Attorney General Vanita Gupta, head of the Justice Department’s Civil Rights Division and U.S. Attorney Wifredo A. Ferrer of the Southern District of Florida. The settlement, which was approved by Miami’s city commission today and will go into effect when the agreement is signed by all parties, resolves claims stemming from the Justice Department’s investigation into officer-involved shootings by MPD officers, which was conducted under the Violent Crime Control and Law Enforcement Act of 1994.  The investigation’s findings, issued in July 2013, identified a pattern or practice of excessive use of force through officer-involved shootings in violation of the Fourth Amendment of the Constitution.  The city’s compliance with the settlement will be monitored by an independent reviewer, former Tampa, Florida, Police Chief Jane Castor.  Under the settlement agreement,

## 2.2 Create a document-term matrix from the preprocessed press releases and to explore top words (5 points)

A. Use the `create_dtm` function I provide (alternately, feel free to write your own!) and create a document-term matrix using the preprocessed press releases; make sure metadata contains the following columns: `id`, `compound` sentiment column you added, and the `topics_clean` column

B. Print the top 10 words for press releases with compound sentiment in the top 5% (so the most positive sentiment)

C. Print the top 10 words for press releases with compound sentiment in the bottom 5% (so the most negative sentiment)

**Hint**: for these, remember the pandas quantile function from pset one.  

D. Print the top 10 words for press releases in each of the three `topics_clean`

For steps B - D, to receive full credit, write a function `get_topwords` that helps you avoid duplicated code when you find top words for the different subsets of the data. There are different ways to structure it but one way is to feed it subsetted data (so data subsetted to one topic etc.) and for it to get the top words for that subset.


In [67]:
def create_dtm(list_of_strings, metadata):
    vectorizer = CountVectorizer(lowercase=True)
    dtm_sparse = vectorizer.fit_transform(list_of_strings)
    dtm_dense_named = pd.DataFrame(dtm_sparse.todense(), 
                                   columns=vectorizer.get_feature_names_out()) 
    metadata_clean = metadata.copy()
    if 'index' in metadata_clean.columns:
        metadata_clean = metadata_clean.drop(columns=['index'])
    dtm_dense_named_withid = pd.concat([metadata_clean.reset_index(drop=True), dtm_dense_named], axis=1)
    dtm_dense_named_withid = dtm_dense_named_withid.loc[:, ~dtm_dense_named_withid.columns.duplicated()]
    return dtm_dense_named_withid

In [68]:
# 2.2 Create a document-term matrix and get top words

# A.
metadata = doj_subset_wscore[["id", "compound", "topics_clean"]]
dtm = create_dtm(doj_subset_wscore["processed_text"], metadata)

# B.
def get_topwords(subset_df, top_n=10):
    word_data = subset_df.drop(columns=["id", "compound", "topics_clean"], errors="ignore")
    return word_data.sum().sort_values(ascending=False).head(top_n)

# C.
top_5 = dtm[dtm["compound"] >= dtm["compound"].quantile(0.95)]
print("Top 10 words in top 5% most positive sentiment:")
print(get_topwords(top_5))

# D.
bottom_5 = dtm[dtm["compound"] <= dtm["compound"].quantile(0.05)]
print("\nTop 10 words in bottom 5% most negative sentiment:")
print(get_topwords(bottom_5))

# E.
for topic in dtm["topics_clean"].unique():
    topic_subset = dtm[dtm["topics_clean"] == topic]
    print(f"\nTop 10 words for topic: {topic}")
    print(get_topwords(topic_subset))

Top 10 words in top 5% most positive sentiment:
agreement     170
enforc        130
state         123
ensur         112
disabl        107
communiti      98
servic         90
court          88
settlement     86
student        85
dtype: int64

Top 10 words in bottom 5% most negative sentiment:
assault     181
crime       179
victim      164
offic       155
hate        137
defend      124
sentenc     104
conspir     104
feder       100
american     95
dtype: int64

Top 10 words for topic: Civil Rights
offic        637
hous         633
discrimin    616
enforc       544
disabl       532
said         497
feder        479
violat       477
state        452
court        414
dtype: int64

Top 10 words for topic: Project Safe Childhood
child          1022
exploit         701
sexual          572
safe            479
childhood       474
project         472
pornographi     452
children        423
crimin          405
prosecut        374
dtype: int64

Top 10 words for topic: Hate Crimes
victim      591

## 2.3 Estimate a topic model using those preprocessed words (5 points)

A. Going back to the preprocessed words from part 2.3.1, estimate a topic model with 3 topics, since you want to see if the unsupervised topic models recover different themes for each of the three manually-labeled areas (civil rights; hate crimes; project safe childhood). You have free rein over the other topic model parameters beyond the number of topics.

B. After estimating the topic model, print the top 15 words in each topic.

**Hints and Resources**:

- Same topic modeling resources linked to above
- Make sure to use the `random_state` argument within the model so that the numbering of topics does not move around between runs of your code

In [71]:
# your code here
from gensim import corpora, models

target_labels = ["civil rights", "hate crimes", "project safe childhood"]
regex_filter = "|".join(target_labels)
doj_filtered = doj_subset_wscore[doj_subset_wscore["topics_clean"].str.contains(regex_filter, case=False, regex=True)].copy()

print(f"{len(doj_filtered)} documents matched the topic filter.")

token_lists = doj_filtered["processed_text"].apply(lambda x: x.split())
token_lists = [tokens for tokens in token_lists if len(tokens) > 0]

if not token_lists:
    print("No documents with usable tokens.")
else:
    dictionary = corpora.Dictionary(token_lists)
    corpus = [dictionary.doc2bow(tokens) for tokens in token_lists]

    lda_model = models.LdaModel(
        corpus=corpus,
        id2word=dictionary,
        num_topics=3,
        random_state=42,
        passes=10
    )

    topic_names = {
        0: "project safe childhood",
        1: "hate crimes",
        2: "civil rights"
    }

    for idx, topic in lda_model.show_topics(num_topics=3, num_words=15, formatted=False):
        words_only = [word for word, _ in topic]
        print(f"\n{topic_names[idx].title()}:\n" + ", ".join(words_only))


717 documents matched the topic filter.

Project Safe Childhood:
child, exploit, sexual, prosecut, safe, sentenc, childhood, project, victim, crimin, pornographi, children, year, guilti, investig

Hate Crimes:
victim, charg, feder, crime, defend, sentenc, prosecut, said, hate, offic, guilti, indict, american, assault, violat

Civil Rights:
hous, discrimin, disabl, enforc, agreement, state, court, said, alleg, settlement, requir, fair, inform, general, provid


## 2.4 Add topics back to main data and explore correlation between manual labels and our estimated topics (10 points)

A. Extract the document-level topic probabilities. Within `get_document_topics`, use the argument `minimum_probability` = 0 to make sure all 3 topic probabilities are returned. Write an assert statement to make sure the length of the list is equal to the number of rows in the `doj_subset_wscores` dataframe

B. Add the topic probabilities to the `doj_subset_wscores` dataframe as columns and create a column, `top_topic`, that reflects each document to its highest-probability topic (eg topic 1, 2, or 3)

C. For each of the manual labels in `topics_clean` (Hate Crime, Civil Rights, Project Safe Childhood), print the breakdown of the % of documents with each top topic (so, for instance, Hate Crime has 246 documents-- if 123 of those documents are coded to topic_1, that would be 50%; and so on). **Hint**: pd.crosstab and normalize may be helpful: https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.crosstab.html

D. Using a couple press releases as examples, write a 1-2 sentence interpretation of why some of the manual topics map on more cleanly to an estimated topic than other manual topic(s)


In [151]:
from sklearn.decomposition import LatentDirichletAllocation

vectorizer = CountVectorizer()
dtm_lda = vectorizer.fit_transform(doj_subset_wscore["processed_text"])

lda_model = LatentDirichletAllocation(n_components=3, random_state=42)
lda_model.fit(dtm_lda)

words = vectorizer.get_feature_names_out()

doc_topics = lda_model.transform(dtm_lda)

assert len(doc_topics) == len(doj_subset_wscore)

LatentDirichletAllocation(n_components=3, random_state=42)

In [152]:
for i in range(doc_topics.shape[1]):
    doj_subset_wscore[f"topic_{i+1}_prob"] = doc_topics[:, i]

doj_subset_wscore["top_topic"] = doc_topics.argmax(axis=1) + 1

In [153]:
topic_summary = pd.crosstab(doj_subset_wscore["topics_clean"],
                            doj_subset_wscore["top_topic"],
                            normalize='index')

topic_summary

top_topic                      1         2         3
topics_clean                                        
Civil Rights            0.393443  0.504918  0.101639
Hate Crimes             0.808943  0.000000  0.191057
Project Safe Childhood  0.006024  0.000000  0.993976

## PART D ANSWER: We notice that topics like Project Safe Childhood can map most accurately maybe because terms used for Project Safe Childhood are much more distinct and consistent whereas topics like Civil Rights and Hate Crimes may have more overlap of terms used from shared legal terms or concepts, thus making them somewhat harder to distinguish. 

# 3. Extend the analysis from unigrams to bigrams (10 points)

In the previous question, you found top words via a unigram representation of the text. Now, we want to see how those top words change with bigrams (pairs of words)

A. Using the `doj_subset_wscore` data and the `processed_text` column (so the words after stemming/other preprocessing), create a column in the data called `processed_text_bigrams` that combines each consecutive pairs of word into a bigram separated by an underscore. Eg:

"depart reach settlem" would become "depart_reach reach_settlem"

Do this by writing a function `create_bigram_onedoc` that takes in a single `processed_text` string and returns a string with its bigrams structured similarly to above example
 
**Hint**: there are many ways to solve but `zip` may be helpful: https://stackoverflow.com/questions/21303224/iterate-over-all-pairs-of-consecutive-items-in-a-list

B. Print the `id`, `processed_text`, and `processed_text_bigram` columns for press release with id = 16-217

In [79]:
## your code here 

# PART A
def create_bigram_onedoc(processed_text):
    words = processed_text.split()
    bigrams = [f"{w1}_{w2}" for w1, w2 in zip(words, words[1:])]
    return " ".join(bigrams)

doj_subset_wscore["processed_text_bigrams"] = doj_subset_wscore["processed_text"].apply(create_bigram_onedoc)

In [80]:
# PART B
example_id = "16-217"
doj_subset_wscore.loc[doj_subset_wscore["id"] == example_id, ["id", "processed_text", "processed_text_bigrams"]]

id  \
313  16-217   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          processed_text  \
313  reach comprehens settlement agreement citi miami miami polic resolv offic involv shoot offic announc princip deputi general vanita gupta head wifredo ferrer southern florida settlement approv miami citi commiss today effect agreement sign parti resolv claim stem offic involv shoot offic conduct violent crime control enforc find issu juli identifi pattern practic excess forc offic involv shoot violat fourth amend constitut citi complianc settlement monitor independ review former tampa florida polic chief jane castor settlement agreement citi implement comprehens reform ensur constitut polic support public trust settlement agreement design minim offic involv shoot effect quick investig offic involv shoot occur measur includ settlement repres renew commit citi miami chief rodolfo llane provid constitut polic miami resid protect public safeti sustain reform said princip deputi general gupta agreement help strengthen relationship communiti serv improv account offic fire weapon unlaw provid communiti particip enforc agreement today agreement result joint effort citi miami ensur miami polic continu effort make communiti safe protect sacr constitut citizen said ferrer oversight communic agreement seek make perman posit chang former chief orosa chief llane made applaud citi commiss vote settlement agreement build upon import reform implement citi sinc issu find includ conduct attorney staff special litig section southern florida   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

C. Use the create_dtm function and the `processed_text_bigrams` column to create a document-term matrix (`dtm_bigram`) with these bigrams. Keep the following three columns in the data: `id`, `topics_clean`, and `compound` 

D. Print the (1) dimensions of the `dtm` matrix from question 2.2  and (2) the dimensions of the `dtm_bigram` matrix. Comment on why the bigram matrix has more dimensions than the unigram matrix 

E. Find and print the 10 most prevelant bigrams for each of the three topics_clean using the `get_topwords` function from 2.2

In [83]:
# your code here

# PART C

metadata_bigram = doj_subset_wscore[["id", "topics_clean", "compound"]]
dtm_bigram = create_dtm(doj_subset_wscore["processed_text_bigrams"], metadata_bigram)

In [84]:
print("Unigram DTM shape:", dtm.shape)
print("Bigram DTM shape:", dtm_bigram.shape)

print("\nExplanation: The bigram DTM has more columns (features) than the unigram DTM because\n"
      "there are many more possible unique two-word combinations than single words.\n"
      "Even if two documents share unigrams, the order of those words affects bigrams —\n"
      "so the bigram vocabulary is naturally larger.")

Unigram DTM shape: (717, 6868)
Bigram DTM shape: (717, 72720)

Explanation: The bigram DTM has more columns (features) than the unigram DTM because
there are many more possible unique two-word combinations than single words.
Even if two documents share unigrams, the order of those words affects bigrams —
so the bigram vocabulary is naturally larger.


In [85]:
for topic in dtm_bigram["topics_clean"].unique():
    topic_subset = dtm_bigram[dtm_bigram["topics_clean"] == topic]
    print(f"\nTop 10 bigrams for topic: {topic}")
    print(get_topwords(topic_subset))


Top 10 bigrams for topic: Civil Rights
fair_hous         231
deputi_general    221
princip_deputi    221
vanita_gupta      202
gupta_head        200
general_vanita    199
said_princip      186
unit_state        156
nation_origin     143
consent_decre     128
dtype: int64

Top 10 bigrams for topic: Project Safe Childhood
safe_childhood       474
project_safe         472
child_pornographi    450
child_exploit        281
sexual_exploit       223
exploit_children     200
plead_guilti         197
exploit_obscen       176
obscen_section       175
child_sexual         174
dtype: int64

Top 10 bigrams for topic: Hate Crimes
hate_crime          379
african_american    367
plead_guilti        275
year_prison         161
special_agent       124
racial_motiv        114
thoma_perez         111
grand_juri          101
perez_general        95
said_thoma           91
dtype: int64


# 4. Optional extra credit (2 points)

You notice that the pharmaceutical kickbacks press release we analyzed in question 1 was for an indictment, and that in the original data, there's not a clear label for whether a press release outlines an indictment (charging someone with a crime), a conviction (convicting them after that charge either via a settlement or trial), or a sentencing (how many years of prison or supervised release a defendant is sentenced to after their conviction).

You want to see if you can identify pairs of press releases where one press release is from one stage (e.g., indictment) and another is from a different stage (e.g., a sentencing).

You decide that one way to approach is to find the pairwise string similarity between each of the processed press releases in `doj_subset`. There are many ways to do this, so Google for some approaches, focusing on ones that work well for entire documents rather than small strings.

Find the top two pairs (so four press releases total)-- do they seem like different stages of the same crime or just press releases covering similar crimes?

In [87]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

texts = doj_subset["contents"].tolist()
ids = doj_subset["id"].tolist()

vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(texts)

cosine_sim = cosine_similarity(tfidf_matrix)

np.fill_diagonal(cosine_sim, 0)
top_pairs = np.dstack(np.unravel_index(np.argsort(cosine_sim.ravel())[-4:], cosine_sim.shape))[0]

for i in range(0, len(top_pairs), 2):
    idx1, idx2 = top_pairs[i]
    print(f"\nPair {i//2 + 1}:")
    print(f"ID {ids[idx1]}:")
    print(texts[idx1][:500], "...")
    print(f"\nID {ids[idx2]}:")
    print(texts[idx2][:500], "...")
    print(f"\nSimilarity score: {cosine_sim[idx1][idx2]:.4f}")


Pair 1:
ID 17-260:
A Weed, California man pleaded guilty today to conspiracy to produce child pornography for his participation in a website that was operated for the purpose of coercing and enticing minors as young as eight years old to engage in sexually explicit conduct on web camera.   Acting Assistant Attorney General Kenneth A. Blanco of the Justice Department’s Criminal Division; U.S. Attorney Dana J. Boente of the Eastern District of Virginia; and Assistant Director Stephen E. Richardson of the FBI’s Crimi ...

ID 17-635:
A Weed, California man was sentenced to 218 months in prison for conspiracy to produce child pornography based on his participation in a website that was operated for the purpose of coercing and enticing minors as young as eight years old to engage in sexually explicit conduct on web camera. Acting Assistant Attorney General Kenneth A. Blanco of the Justice Department’s Criminal Division; U.S. Attorney Dana J. Boente of the Eastern District of Virginia; and S

In [88]:
# your code here 